crates/ide-assists/src/handlers/auto_import.rs

```diff
@@ -158,11 +165,144 @@ fn group_label(import_candidate: &ImportCandidate) -> GroupLabel {
     GroupLabel(name)
 }
 
+/// Determine how relevant a given import is in the current context. Higher scores are more
+/// relevant.
+fn relevance_score(ctx: &AssistContext, import: &LocatedImport) -> i32 {
+    let mut score = 0;
+
+    let db = ctx.db();
+
+    let item_module = match import.item_to_import {
+        hir::ItemInNs::Types(item) | hir::ItemInNs::Values(item) => item.module(db),
+        hir::ItemInNs::Macros(makro) => Some(makro.module(db)),
+    };
+
+    let current_node = match ctx.covering_element() {
+        NodeOrToken::Node(node) => Some(node),
+        NodeOrToken::Token(token) => token.parent(),
+    };
+
+    if let Some(module) = item_module.as_ref() {
+        if module.krate().is_builtin(db) {
+            // prefer items builtin to the language
+            score += 5;
+        }
+    }
+
+    let current_scope = current_node.as_ref().and_then(|node| ctx.sema.scope(node));
+
+    match item_module.zip(current_scope) {
+        // get the distance between the modules (prefer items that are more local)
+        Some((item_module, current_scope)) => {
+            let current_module = current_scope.module();
+            score -= module_distance_hueristic(&current_module, &item_module, db) as i32;
+        }
+
+        // could not find relevant modules, so just use the length of the path as an estimate
+        None => return -(2 * import.import_path.len() as i32),
+    }
+
+    score
+}
+
+/// A heuristic that gives a higher score to modules that are more separated.
+fn module_distance_hueristic(current: &Module, item: &Module, db: &dyn HirDatabase) -> usize {
+    let mut current_path = current.path_to_root(db);
+    let mut item_path = item.path_to_root(db);
+
+    current_path.reverse();
+    item_path.reverse();
+
+    let mut i = 0;
+    while i < current_path.len() && i < item_path.len() {
+        if current_path[i] == item_path[i] {
+            i += 1
+        } else {
+            break;
+        }
+    }
```

Comment:
This can be done with iterators, the following should be equivalent I believe
```rs
current_path.iter().zip(&item_path).position(|(a, b)| a != b).unwrap_or(0)
```

Reply:
Sure, but the `unwrap_or` should be the shortest of the two paths. I did this using `take_while` instead (8e5b318d99d3d215ab0ae84b1a80e8034975002d).

Reply:
Oh ye good catch